### PR TITLE
test: Randomize the order in which simulated timers scheduled for the same monotonic time execute.

### PR DESCRIPTION
**Pull request by antoniovicente**
Commit Message: test: Randomize the order in which simulated timers scheduled for the same monotonic time execute.
Additional Description: The order in which real timers scheduled for the exact same wall time execute is non-deterministic.  Simulated timer behavior should be as similar as possible as that of real timers.
Risk Level: n/a, test only changes
Testing: unit
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: test/test_common/simulated_time_system.h [hunk 1]
```diff
@@ -105,8 +115,9 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
   RealTimeSource real_time_source_; // Used to initialize monotonic_time_ and system_time_;
   MonotonicTime monotonic_time_ ABSL_GUARDED_BY(mutex_);
   SystemTime system_time_ ABSL_GUARDED_BY(mutex_);
+  Runtime::RandomGeneratorImpl random_source_ ABSL_GUARDED_BY(mutex_);
```
Comment (alyssawilk) [inline test/test_common/simulated_time_system.h:118]: driveby: if you use TestRandom it'll cerr the seed for reproducability and allow flag based override.

Comment (antoniovicente) [reply test/test_common/simulated_time_system.h]: Thanks, done.

**Comment by antoniovicente** [comment]
@jmarantz ping

**Comment by jmarantz** [comment]
I looked a bit deeper and realize this relies on `TestRandomGenerator`, which was pre-existing. I think the big concern that I have is that the methodology around that is not obvious.

If CI fails will a developer know to look for the seed printout? If it fails during development, will a developer capture the seed before overwriting it?

I haven't noticed this as an issue before, maybe because TestRandomGenerator doesn't have a ton of usage yet? But this would give it more usage. Maybe we could add some text to CONTRIBUTING.md or a new `.md` file about debugging/reproducing flakes? That could be a follow-up.

File: test/test_common/simulated_time_system.h [hunk 2]
```diff
@@ -64,10 +65,24 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
   class SimulatedScheduler;
   class Alarm;
   friend class Alarm; // Needed to reference mutex for thread annotations.
-  struct CompareAlarms {
-    bool operator()(const Alarm* a, const Alarm* b) const;
+  struct AlarmRegistration {
+    MonotonicTime time;
+    // Random tie-breaker for alarms scheduled for the same monotonic time used to mimic
+    // non-deterministic execution of real alarms scheduled for the same wall time.
+    uint64_t randomness;
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.h:72]: where does this get initialized?

File: test/test_common/simulated_time_system.cc [hunk 3]
```diff
@@ -308,34 +279,40 @@ Thread::CondVar::WaitStatus SimulatedTimeSystemHelper::waitFor(Thread::MutexBasi
...
-
 void SimulatedTimeSystemHelper::addAlarmLockHeld(
     Alarm* alarm, const std::chrono::microseconds& duration) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   ASSERT(&(alarm->timeSystem()) == this);
-  alarm->setTimeLockHeld(monotonic_time_ + duration);
-  alarms_.insert(alarm);
+  ASSERT(alarms_.size() == alarm_registrations_map_.size());
+  ASSERT(alarm_registrations_map_.find(alarm) == alarm_registrations_map_.end());
+
+  auto insert_result = alarms_.insert({monotonic_time_ + duration, random_source_.random(), alarm});
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.cc:295]: I found it hard to find this initialization line. WDYT of giving AlarmRegistration a ctor:
```
  AlarmRegistration(MonotonicTime time, uint64 randomness, Alarm* alarm) : ...
```
so that it was more obvious from looking at the class how the member variables must always be initialized.

File: test/test_common/simulated_time_system.cc [hunk 4]
```diff
@@ -308,34 +279,40 @@ Thread::CondVar::WaitStatus SimulatedTimeSystemHelper::waitFor(Thread::MutexBasi
...
+  ASSERT(insert_result.second);
+  alarm_registrations_map_.emplace(alarm, insert_result.first);
+
+  // Sanity check that the parallel data structures used for alarm registration have the same number
+  // of entries.
+  ASSERT(alarms_.size() == alarm_registrations_map_.size());
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.cc:301]: I'm a little unclear why the structure of this had to change so much, as opposed to simply replacing `index_` with `randomness_` in the structure.

Does that provide some flexibility previously lacking?

File: test/test_common/simulated_time_system.h [hunk 5]
```diff
@@ -64,10 +65,24 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
   class SimulatedScheduler;
   class Alarm;
   friend class Alarm; // Needed to reference mutex for thread annotations.
-  struct CompareAlarms {
-    bool operator()(const Alarm* a, const Alarm* b) const;
+  struct AlarmRegistration {
+    MonotonicTime time;
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.h:69]: In Envoy style we end struct member variables with `_` just like classes, and unlike native google style.

See https://github.com/envoyproxy/envoy/blob/master/STYLE.md#deviations-from-google-c-style-guidelines

File: test/test_common/simulated_time_system.cc (hunk 4 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.cc:301]: I wanted to recompute randomness on every insert, not just pick a random value at creation time.  Also, having the timeout outside the timer is closer to what I'm used to.

File: test/test_common/simulated_time_system.h (hunk 2 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.h:72]: Seems like you found it.

File: test/test_common/simulated_time_system.cc (hunk 3 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.cc:295]: IIRC there's plenty of prior art in Envoy for use of the c++ 11 struct initializer syntax in these cases.

Added constructor implementation.

File: test/test_common/simulated_time_system.h (hunk 5 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.h]: Done.

**Comment by antoniovicente** [comment]
> I looked a bit deeper and realize this relies on `TestRandomGenerator`, which was pre-existing. I think the big concern that I have is that the methodology around that is not obvious.
> 
> If CI fails will a developer know to look for the seed printout? If it fails during development, will a developer capture the seed before overwriting it?
> 
> I haven't noticed this as an issue before, maybe because TestRandomGenerator doesn't have a ton of usage yet? But this would give it more usage. Maybe we could add some text to CONTRIBUTING.md or a new `.md` file about debugging/reproducing flakes? That could be a follow-up.

I'll look into this a bit later today.  Any other concerns on this PR?

File: test/test_common/simulated_time_system.h [hunk 6]
```diff
@@ -64,10 +65,27 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
...
-    bool operator()(const Alarm* a, const Alarm* b) const;
+  struct AlarmRegistration {
+    AlarmRegistration(MonotonicTime time, uint64_t randomness, Alarm* alarm)
+        : time_(time), randomness_(randomness), alarm_(alarm) {}
+
+    MonotonicTime time_;
+    // Random tie-breaker for alarms scheduled for the same monotonic time used to mimic
+    // non-deterministic execution of real alarms scheduled for the same wall time.
+    uint64_t randomness_;
+    Alarm* alarm_;
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.h:76]: it would be more Envoyish to use `Alarm&` here, as it can't be null.

File: test/test_common/simulated_time_system.h [hunk 7]
```diff
@@ -64,10 +65,27 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
...
+
+    friend bool operator<(const AlarmRegistration& lhs, const AlarmRegistration& rhs) {
+      if (lhs.time_ != rhs.time_) {
+        return lhs.time_ < rhs.time_;
+      }
+      if (lhs.randomness_ != rhs.randomness_) {
+        return lhs.randomness_ < rhs.randomness_;
+      }
+      return lhs.alarm_ < rhs.alarm_;
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.h:85]: you are intentionally comparing pointers here?

File: test/test_common/simulated_time_system.cc (hunk 4 above)
Comment (jmarantz) [reply test/test_common/simulated_time_system.cc:301]: That doesn't seem at all hard to do in the current structure; we need to know if an alarm is in the map, and can't mutate a map-key while in the map. I think this is just a question of mutating the randomness_ field immediately prior to insertion, which happens only in `addAlarmLockHeld`.

I'm not that wedded to the structure that is there now, but it has these 2 benefits:
 * easier to review due to less changes in code
 * not having 2 parallel structures to keep in sync.

File: test/test_common/simulated_time_system.h (hunk 7 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.h:85]: yes.  This is a final tie breaker that is almost impossible to hit.

Comment (jmarantz) [reply test/test_common/simulated_time_system.h:85]: I see...it would be hit only if you get two of the same random number. Worth a comment?

File: test/test_common/simulated_time_system.h [hunk 8]
```diff
@@ -64,10 +65,27 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
...
+
+    MonotonicTime time_;
+    // Random tie-breaker for alarms scheduled for the same monotonic time used to mimic
+    // non-deterministic execution of real alarms scheduled for the same wall time.
+    uint64_t randomness_;
+    Alarm* alarm_;
+
+    friend bool operator<(const AlarmRegistration& lhs, const AlarmRegistration& rhs) {
+      if (lhs.time_ != rhs.time_) {
+        return lhs.time_ < rhs.time_;
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.h:80]: if lhs.time_ > rhs.time_ shouldn't we return false here? We don't want to then fall thru to `.randomness` then right?

File: test/test_common/simulated_time_system.h (hunk 7 above)
Comment (jmarantz) [reply test/test_common/simulated_time_system.h:85]: Also, while this is super-rare, it introduces non-determinism outside the scope of the `--seed` flag. You could also retain `index_` but just have `randomness_` supersede.

that would also allow users to specify `--seed 0` and disable the non-determinism, in case someone wants that while debugging. For seed==0 we'd just populate `.randomness=0` and fall-through to index.

File: test/test_common/simulated_time_system.h (hunk 8 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.h:80]: We only fallback to randomness when the two times are equal.  The code already behaves as you suggest.

File: test/test_common/simulated_time_system.cc (hunk 4 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.cc:301]: I'm really not a fan of the current structure.  Mutating the alarm feels too subtle, brittle and error prone.  I can offer to do a no-op refactor to an alternate structure first.

File: test/test_common/simulated_time_system.h (hunk 7 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.h:85]: --seed=0 already means "use a random seed".  The odds of randomness_ being equal are close to 1 in 2**64 since the number of alarms used in a test is relatively small.

File: test/test_common/simulated_time_system.cc (hunk 4 above)
Comment (jmarantz) [reply test/test_common/simulated_time_system.cc:301]: Actually if you are signing up to own this I am totally down with the rewrite :)

No need for an intermediate refactor.

File: test/test_common/simulated_time_system.h (hunk 7 above)
Comment (jmarantz) [reply test/test_common/simulated_time_system.h:85]: Yeah my argument was pedantic; this is OK but IMO worth a comment.

File: test/test_common/simulated_time_system.h (hunk 8 above)
Comment (jmarantz) [reply test/test_common/simulated_time_system.h:80]: indeed; my bad; this logic is fine.

File: test/test_common/simulated_time_system.h [hunk 9]
```diff
@@ -120,7 +124,8 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
   SystemTime system_time_ ABSL_GUARDED_BY(mutex_);
   TestRandomGenerator random_source_ ABSL_GUARDED_BY(mutex_);
   AlarmSet alarms_ ABSL_GUARDED_BY(mutex_);
-  std::map<Alarm*, AlarmSet::const_iterator> alarm_registrations_map_ ABSL_GUARDED_BY(mutex_);
+  std::unordered_map<Alarm*, AlarmSet::const_iterator>
```
Comment (jmarantz) [inline test/test_common/simulated_time_system.h:127]: any reason not to use absl::flat_hash_map?

Comment (antoniovicente) [reply test/test_common/simulated_time_system.h]: Done.

File: test/test_common/simulated_time_system.h (hunk 7 above)
Comment (antoniovicente) [reply test/test_common/simulated_time_system.h]: Added comment and ASSERT.
